debug/satellites-custom-layer.js

```diff
@@ -0,0 +1,176 @@
+    gl.attachShader(program, vertShader);
+    gl.attachShader(program, fragShader);
+    gl.linkProgram(program);
+    gl.validateProgram(program);
+
+    if (!gl.getProgramParameter(program, gl.LINK_STATUS)) {
+        const info = gl.getProgramInfoLog(program);
+        console.error(`Could not compile WebGL program. \n\n${info}`);
+    }
+
+    return program;
+};
+
+function updateVboAndActivateAttrib(gl, prog, vbo, data, attribName) {
+    gl.bindBuffer(gl.ARRAY_BUFFER, vbo);
+    gl.bufferData(gl.ARRAY_BUFFER, new Float32Array(data), gl.DYNAMIC_DRAW);
+    const attribLoc = gl.getAttribLocation(prog, attribName);
+    gl.vertexAttribPointer(attribLoc, 3, gl.FLOAT, false, 0, 0);
+    gl.enableVertexAttribArray(attribLoc);
+}
+
+const satellitesLayer = {
+    id: 'satellites',
+    type: 'custom',
+    onAdd (map, gl) {
+        this.map = map;
+
+        this.posEcef = [];
+        this.posMerc = [];
+
+        this.posEcefVbo = gl.createBuffer();
+        this.posMercVbo = gl.createBuffer();
+
+        this.globeProgram = createProgram(gl, globeVertCode, fragCode);
+        this.mercProgram = createProgram(gl, mercVertCode, fragCode);
+
+        fetch('space-track-leo.txt').then(r => r.text()).then(rawData => {
+            const tleData = rawData.replace(/\r/g, '')
+              .split(/\n(?=[^12])/)
+              .filter(d => d)
+              .map(tle => tle.split('\n'));
+            this.satData = tleData.map(([name, ...tle]) => ({
+              satrec: satellite.twoline2satrec(...tle),
+              name: name.trim().replace(/^0 /, '')
+            }))
+            // exclude those that can't be propagated
+            .filter(d => !!satellite.propagate(d.satrec, new Date()).position)
+            .slice(0, 10);
+
+            this.updateBuffers();
+        });
+    },
+
+    updateBuffers() {
+        time = new Date(+time + TIME_STEP);
+        const gmst = satellite.gstime(time);
+        this.posEcef = [];
+        this.posMerc = [];
+        for (let i = 0; i < this.satData.length; ++i) {
+            const satrec = this.satData[i].satrec;
+            const eci = satellite.propagate(satrec, time);
+            if (eci.position) {
+                const geodetic = satellite.eciToGeodetic(eci.position, gmst);
+
+                const lngLat = [satellite.degreesLong(geodetic.longitude), satellite.degreesLat(geodetic.latitude)];
+                const altitude = geodetic.height * KM_TO_M;
+
+                const merc = mapboxgl.MercatorCoordinate.fromLngLat(lngLat, altitude);
+                const ecef = mapboxgl.LngLat.convert(lngLat).toEcef(altitude);
+
+                this.posEcef.push(...ecef);
+                this.posMerc.push(...[merc.x, merc.y, merc.z]);
+            }
+        }
+    },
+
+    render (gl, projectionMatrix, globeMatrix) {
+        if (this.satData) {
+            this.updateBuffers();
+
+            gl.enable(gl.DEPTH_TEST);
+            if (globeMatrix !== null) { // globe projection
```

Comment:
For discussion: I wonder if there's a better way on the API of custom layer to split between projections. I could imagine two more options:
- add a specific value to express which projection we're in (e.g. `render(gl, projection, matrices, ...)`
- have another call for this projection so related code is contained within, e.g. `renderGlobe`

I find it a bit counter-intuitive to rely on one of the input parameter not being set, but also a bit fragile, if we internally change the interface of this function to not return `null`, https://github.com/mapbox/mapbox-gl-js/blob/ad9c713939882f23db7902361d1d34a96280def6/src/geo/transform.js#L1652 we would break such an API and also client-side code.

If you still prefer that design, I'd only suggest to ensure that function always return `null` under these circumstances with some small tests covering that.

Reply:
@karimnaaji that's a very good point! There was an implicit agreement with @astojilj that we could always get back to these (i.e. immediate mode related APIs) later as they're experimental anyways. I think we could go with the approach you suggested.

Reply:
`renderGlobe` seems better. There is a need to pass additional information, globe matrix, mercatorMatrix (projection matrix passed during globe projection has different z scale from mercator projection matrix used in mercator projection, due to pixelsPerMeter usage).

https://github.com/mapbox/mapbox-gl-js/blob/55eab7cf381723e6a3a055260771c8e26f6628d6/src/geo/transform.js#L1874-L1875